agent/run_job.go

```diff
@@ -0,0 +1,236 @@
+			r.logStreamer.Process([]byte(
+				"Some containers had unknown exit statuses. Perhaps they were in ImagePullBackOff.",
+			))
+		}
+	}
+
+	// Add the final output to the streamer
+	r.logStreamer.Process(r.output.ReadAndTruncate())
+
+	// Collect the finished process' exit status
+	exit.Status = fmt.Sprintf("%d", r.process.WaitStatus().ExitStatus())
+
+	if ws := r.process.WaitStatus(); ws.Signaled() {
+		exit.Signal = process.SignalString(ws.Signal())
+	}
+
+	if r.stopped {
+		// The agent is being gracefully stopped, and we signaled the job to end. Often due
+		// to pending host shutdown or EC2 spot instance termination
+		exit.SignalReason = "agent_stop"
+	} else if r.cancelled {
+		// The job was signaled because it was cancelled via the buildkite web UI
+		exit.SignalReason = "cancel"
+	}
+
+	return exit
+}
+
+func (r *JobRunner) cleanup(ctx context.Context, wg *sync.WaitGroup, exit *processExit) {
+	finishedAt := time.Now()
+
+	// Stop the header time streamer. This will block until all the chunks have been uploaded
+	r.headerTimesStreamer.Stop()
+
+	// Stop the log streamer. This will block until all the chunks have been uploaded
+	r.logStreamer.Stop()
+
+	// Warn about failed chunks
+	if count := r.logStreamer.FailedChunks(); count > 0 {
+		r.logger.Warn("%d chunks failed to upload for this job", count)
+	}
+
+	// Wait for the routines that we spun up to finish
+	r.logger.Debug("[JobRunner] Waiting for all other routines to finish")
+	wg.Wait()
+
+	// Remove the env file, if any
+	if r.envFile != nil {
+		if err := os.Remove(r.envFile.Name()); err != nil {
+			r.logger.Warn("[JobRunner] Error cleaning up env file: %s", err)
+		}
+		r.logger.Debug("[JobRunner] Deleted env file: %s", r.envFile.Name())
+	}
+
+	// Write some metrics about the job run
+	jobMetrics := r.conf.MetricsScope.With(metrics.Tags{"exit_code": exit.Status})
+
+	if exit.Status == "0" {
+		jobMetrics.Timing("jobs.duration.success", finishedAt.Sub(r.startedAt))
+		jobMetrics.Count("jobs.success", 1)
+	} else {
+		jobMetrics.Timing("jobs.duration.error", finishedAt.Sub(r.startedAt))
+		jobMetrics.Count("jobs.failed", 1)
+	}
+
+	// Finish the build in the Buildkite Agent API
+	// Once we tell the API we're finished it might assign us new work, so make sure everything else is done first.
+	r.finishJob(ctx, finishedAt, exit, r.logStreamer.FailedChunks())
+
+	r.logger.Info("Finished job %s", r.conf.Job.ID)
+
+}
+
+func (r *JobRunner) CancelAndStop() error {
+	r.cancelLock.Lock()
+	r.stopped = true
+	r.cancelLock.Unlock()
+	return r.Cancel()
+}
+
+func (r *JobRunner) Cancel() error {
+	r.cancelLock.Lock()
+	defer r.cancelLock.Unlock()
+
+	if r.cancelled {
+		return nil
+	}
+
+	if r.process == nil {
+		r.logger.Error("No process to kill")
+		return nil
+	}
+
+	reason := ""
+	if r.stopped {
+		reason = " (agent stopping)"
+	}
+	r.logger.Info("Canceling job %s with a grace period of %ds%s",
+		r.conf.Job.ID, r.conf.AgentConfiguration.CancelGracePeriod, reason)
```

Comment:
It's so strange that reason starts with a space. If you can find out why from the git blame, can you surface that as a comment?

Reply:
afaict, `reason` is never used anywhere other than that log line below, where it looks like it has a space prepended so that the log line doesn't have a trailing space? very odd

Reply:
~~Kill~~ trim it!

Reply:
i updated it so that it just has a trailing space if there's no `reason` - i think readability trumps whatever reason we have to not have trailing spaces in log lines